docs/v1-proposed-api.md

```diff
@@ -0,0 +1,295 @@
+  namespace: uhc-staging-18urm2sbab85d9crlk27is39n706ph3g
+spec:
+  name: worker
+  platform:
+    aws:
+      rootVolume:
+        iops: 100
+        size: 300
+        type: gp2
+      type: m5.xlarge
+      zones:
+      - us-east-1a
+      - us-east-2a
+  replicas: 4
+  # future addition, would not be used in conjunction with replicas:
+  autoscaling:
+    minReplicas: 5
+    maxReplicas: 10
+```
+
+## ClusterDeprovisionRequest
+
+  1. Rename to ClusterDeprovision, to better match ClusterProvision.
+
+## SyncSet + SelectorSyncSet
+
+  1. Drop support for ApplyOnce patch mode.
+
+## HiveConfig
+
+  1. Adjust HiveConfig API to map managed DNS domains to cloud provider they should be used for.
+    * Would allow using Hive to provision to multiple clouds, using that cloud for DNS forwarding, from one Hive.
+    * This is not the expected deployment model in Dedicated (one Hive per cloud)
+    * Convert managedDomains to a list of structs rather than just a string. Struct should include a list of the platforms we will manage DNS for on this sub-domain.
+    * Could have multiple domains on AWS using different certs.
+  1. Rename ExternalDNS to ManageDNS (section where the credentials are stored, ExternalDNS being phased out)
+
+
+# Migrating to V1
+
+CRD versioning functionality is not available in OpenShift 3.11 where prod/stage run and will continue to do so until private cluster support is ready and we can migrate. We must however move to V1 as soon as possible.
+
+To avoid having to fork and abandon stage/prod until they can move to 4.3+ we propose the following solution using an aggregated API server.
+
+Migration will be launched manually by SRE (with Hive team present), logged into the cluster as cluster admin. Migration will proceed in phases, so we can examine state of cluster to ensure everything is progressing ok.
+
+## Prerequisites
+
+  1. Deploy a version of Hive with twiest's DR owner reference restoration labelling. We will need these labels to restore owner refs.
+
+## Migration Steps
+
+  1. Phase 1
+    1. Scale down hive-operator, hive-controllers, and wait until no pods are running.
+    1. Fetch yaml for all Hive v1alpha1 CRs and store on disk.
+    1. Remove all v1alpha1 finalizers.
+  1. Phase 2
+    1. Ensure no hive pods running.
+    1. Delete all v1alpha1 CRs. Seems safer than orphaning them in etcd and just deleting the CRDs.
+    1. Delete all v1alpha1 CRDs.
+  1. Phase 3
+    1. Deploy new version of Hive with v1 CRDs and v1alpha1 aggregated API by merging v1 branch into master.
+      * Wait for it to be up and running. No v1 CRDs will exist
+    1. Scale down new pods for hive-operator and hive-controllers again, but leave the v1alpha1 aggregated API running.
+    1. oc apply the backups of v1alpha1 CRs and let the aggregated API server migrate them and store them as v1.
+    1. Scale up hive-operator and hive-controllers, allow them to restore owner refs for us as if this was a DR restore.
```

Comment:
@jharrington22 could you ensure the right people have seen the above? I realize we may have some problems here.

At one point during this migration we're going to be removing all finalizers from our Hive types, and then deleting the CRs. Presumably this is going to cause problems for SRE operators which likely do not restore them, they would try to act again when the new migrated types are created wouldn't they.

Reply:
Maybe we need to expand on this a little.

Ensure SRE operators are scaled down as well, and explicitly restore all finalizers as part of the migration, before we scale everything back up.